File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/core/signature/RFC2616DatedSigner.java
package uk.nhs.nhsx.core.signature;

import java.time.Instant;
import java.time.ZoneId;
import java.time.format.DateTimeFormatter;
import java.util.Locale;
import java.util.function.Function;
import java.util.function.Supplier;

public class RFC2616DatedSigner implements DatedSigner {

    private static final DateTimeFormatter format = DateTimeFormatter.ofPattern("EEE, dd MMM yyyy HH:mm:ss zzz", Locale.ENGLISH);
    

    private final Supplier<Instant> clock;
    private final Signer signer;

    public RFC2616DatedSigner(Supplier<Instant> clock, Signer signer) {
        this.clock = clock;
        this.signer = signer;
    }

    @Override
    public DatedSignature sign(Function<DatedSignature.SignatureDate, byte[]> content) {

        Instant instant = clock.get();
        String date = format.format(instant.atZone(ZoneId.of("UTC")));

        DatedSignature.SignatureDate signatureDate = new DatedSignature.SignatureDate(date, instant);

        return new DatedSignature(signatureDate,signer.sign(content.apply(signatureDate)));
    }
}

File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/core/signature/RFC2616DatedSigner.kt
package uk.nhs.nhsx.core.signature

import uk.nhs.nhsx.core.Clock
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.Locale

class RFC2616DatedSigner(private val clock: Clock, private val signer: Signer) : DatedSigner {
    private val format = DateTimeFormatter.ofPattern("EEE, dd MMM yyyy HH:mm:ss zzz", Locale.ENGLISH)

    override fun sign(content: (SignatureDate) -> ByteArray): DatedSignature {
        val instant = clock()
        val date = format.format(instant.atZone(ZoneId.of("UTC")))
        val signatureDate = SignatureDate(date, instant)
        return DatedSignature(signatureDate, signer.sign(content(signatureDate)))
    }
}
